ISSUE-20: Webforms as input method for Archipelago Digital Objects (ADO)

DiegoPino opened the pull request:
#What is new here?
See #20 
First attempt at documenting what Display Modes/View Modes/Form Modes are and how to connect a SBF widget that uses Webforms as processing to a custom webform.

# What do i need?

Proofreading, helping make this clearer and hopefully this will also speak to some UI needs we want to address  @giancarlobi @mitchellkeaney @marlo-longley

Will share on the Groups for a first pass

DiegoPino committed:
ebforms as input method for Archipelago Digital Objects (ADO)

First attempt at documenting what Display Modes/View Modes/Form Modes are and how to connect a SBF widget that uses Webforms as processing to a custom webform.

DiegoPino committed:
Gosh, wrong location!

DiegoPino committed:
make broken image paths relative

giancarlobi commented on webformsasinput.md:
```diff
@@ -0,0 +1,133 @@
+You can see that for every existint Content Type, there is a drop down menu with options:
+
+- Manage Display: will lead you to configuration page where you can setup each View Mode and its settings for a given Content Type
+- Form Display: will lead you to configuration page where you can setup each View Mode and its settings for a given Content Type
```
@DiegoPino I think here we have to change to:
Manage Form Display: will lead you to configuration page where you can setup each **Form** Mode and its settings for a given Content Type

DiegoPino replied:
True, thanks! Will correct

giancarlobi commented on webformsasinput.md:
```diff
@@ -0,0 +1,133 @@
+The Table that follows has one row per Field attached/partof this Content Type. The list of fields here is shorter, the SBF CopyFields are not present because all data goes really only into real fields. Also some other, display only ones (means you can not modify them) will not appear here. Again, Some of the fields are part of the Content Type itself, in this case Digital Object (bundled) and so ...
+"Field" column contains each field name and the Widget Colum allows you to select what type of Input you are going to use to feed it on Ingest/edit. On the right you will see again a little gear, that allows you to configure the settings for a particular Widget. Those settings apply always only to the current Form Mode.
+
+So. The one we want to understand is the one attached to the "Descriptive Metadata" field. Currently one named "Strawberryfield webform basewd input with inline rendering". There are other two. but lets start with this. Press on the Gear to the right on the same row.
```
@DiegoPino a typo here "Strawberryfield webform **based** input with inline rendering"

giancarlobi commented on webformsasinput.md:
```diff
@@ -0,0 +1,133 @@
+![Webform General Settings](../imgs/webform-general-settings.jpg)
+Gist here is (look at the screenshot and copy the settings):
+
+- Disable Saving of Form Submissions. You won't need this form to generate a Native Webform Submission entry. 
```
@DiegoPino Here it's not clear to me if we have to check or uncheck the option

DiegoPino replied:
Ok, will reword :)

giancarlobi commented on webformsasinput.md:
```diff
@@ -0,0 +1,133 @@
+
+Save your settings. And you are ready to roll. That webform can now be used as a Setting for any of the StrawberryField Widgets that use Webforms.
+
+Finally (the real finally). Archipelago encourages at least one Field/JSON key to be present always. The "type" JSON key. So make sure that your Custom Webform has that one, you can copy its config from the provided Webform (in the future we will generate it automatically for you, no worries, we are working to make this easier). having a "type" value will make your life easier. You don't need it, ...
```
@DiegoPino Do we need here to specify where check if type JSON key is present? Not so clear to me.

DiegoPino replied:
Also true. I was almost giving up on those last sentences. I feel setting up/creating a webform should almost its own documentation. I will add more details during the morning. Now its a good time to retake your idea of a repository of example webforms. Feels setting some of these things directly on the yaml is faster. Also. I want to code some of these into the handler. Basically settings we need will be forced by the handler. Anyway, you are right !

giancarlobi reviewed (approved):
@DiegoPino What a great work!!! I left some minor notes. For me is ok but I'm not English mother toungue so other have to check syntax. Thanks again for this.

DiegoPino committed:
Address Review. Typo

DiegoPino committed:
Address more code review notes

DiegoPino commented:
@giancarlobi thanks! And thanks for reading this. Its a lot! I think i addressed all code reviews. Will recheck later today. Cheers!

giancarlobi commented:
@DiegoPino The hard work is write it not reading so thanks to you for this big effort.

DiegoPino committed:
Small wording improvement

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+- Widgets do one thing right: they expose some type of Form/UI interaction that allows a user to input data into the Entity, under that specific field. And of course they make sure that what you input is validated and saved (if good) correctly.
+- Which Widgets are available will depend on the "type" of field the Content Entity has. 
+  - Example: A `Node` Title will have a single Text Input with some options, like the size of the Textfield used to feed it.
+  - More Complex and double fun fields, like the ones of type `SBF`, will provide a larger list of possible Widgets, randing from raw JSON input (like what you would do if your data was already in the right format) to the reason we are reading this: `Webform driven Widgets` like the ones the _webform_strawberrfield_ Drupal module provides, ones that use an existing Webform (which are also Entitit ...
```
```suggestion
  - More Complex and fun fields, like the ones of type `SBF` (strawberryfield), will provide a larger list of possible Widgets, ranging from raw JSON input (which you could select if your data was already in the right format) to the reason we are reading this: `Webform driven Widgets`. These Widgets include:
   - ones the _webform_strawberrfield_ Drupal module provides
   - ones that use an existing Webform (which are also Entitites!) which either 1) you created or 2) we provided as a setting
   
If you chose a widget other than the raw JSON, the widget will take the raw JSON to build, massage and enrich the data so that it can be presented in a visual format by the SBF. This is because a SBF type of field has much more than just a text value. It contains a full graph of metadata and properties, inclusive links to Files and provenance metadata, which for example allows us to use an Upload field directly in the attached/configured webform.
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+
+The main difference, other than their purpose (Output v/s Input) is that, on View Modes, the settings you apply to each field are associated to "Formatters" and on Form Modes, the settings you apply to each field are connected to "Widgets".
+
+So, resuming, what lives under the Concept of a "Display Mode": 
```
Very picky. This would be correct, but anybody would understand as it is now.

So, resuming, what lives under the Concept of a "Display Mode": 
```suggestion
So, resuming, this is what lives under the Concept of a "Display Mode": 
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+
+# I think i get this...but how can i use this knowledge now?
+
+Good question! So, to enable, configure and customize these Display Modes you have to navigate go to your `Content Type` Configuration page, in your running Archipelago, which can be found at `/admin/structure/types`. Note: the way things are named in Drupal can be confusing to even the most deeply committed Drupal user, so bear in mind some terms will change. Feel free to read and re-read.
```
```suggestion
Good question! So, to enable, configure, and customize these Display Modes you have to navigate to your `Content Type` Configuration page in your running Archipelago. This is found at `/admin/structure/types`. Note: the way things are named in Drupal can be confusing to even the most deeply committed Drupal user, so bear in mind some terms will change. Feel free to read and re-read.
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+
+![Display Mode Managment for Content Types](../imgs/managing-display-modes.jpg)
+
+You can see that for every existint Content Type, there is a drop down menu with options:
```
```suggestion
You can see that for every existent Content Type, there is a drop down menu with options:
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+![Manage Display](../imgs/manage-display.jpg)
+
+On the top you will see all your View Modes Listed, with the `Default` one selected and expanded.
+The Table that follows has one row per Field attached/partof this Content Type. Some of the fields are part of the Content Type itself, in this case Digital Object (bundled) and some other ones are common to every Content Entity derived from a Node.
```
```suggestion
The Table that follows has one row per Field attached/part of this Content Type. Some of the fields are part of the Content Type itself, in this case Digital Object (bundled) and some other ones are common to every Content Entity derived from a Node.
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+
+On the top you will see all your View Modes Listed, with the `Default` one selected and expanded.
+The Table that follows has one row per Field attached/partof this Content Type. Some of the fields are part of the Content Type itself, in this case Digital Object (bundled) and some other ones are common to every Content Entity derived from a Node.
+"Field" column contains each field name (not their type, reason why you don't see Strawberry Field there!) but we can tell your right now that there is one, named "Descriptive Metadata" that is of `SBF` type. 
```
```suggestion
"Field" column contains each field name (not their type, reason why you don't see Strawberry Field there!) but we can tell you right now that there is one, named "Descriptive Metadata" that is of `SBF` type. 
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+
+*Also Surprise:* You Content Entity has really really just 2 fields! And that, friends, is one of the secret ingredients of Archipelago. All goes into a Single Field.
+But wait: i see more fields in my Manage Display table. Why?
+Well. Some of them are base fields, part of what a Drupal Node is: base field means you can not remove them, they are part of the Defintion itself. One obvious one is the `Title`. 
```
```suggestion
Well. Some of them are base fields, part of what a Drupal Node is: base field means you can not remove them, they are part of the Definition itself. One obvious one is the `Title`. 
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+But wait: i see more fields in my Manage Display table. Why?
+Well. Some of them are base fields, part of what a Drupal Node is: base field means you can not remove them, they are part of the Defintion itself. One obvious one is the `Title`. 
+
+But there are also some very particular to Archipelago ones: You can see there are also ones named
```
```suggestion
But there are also some fields very particular to Archipelago: You can see there are also ones named
```

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,156 @@
+"Formatter Object Metadata", "Media" and one named "Static Media"!. Where does come from? Those are also Strawberryfields. It sounds confusing but it is really simple. They are really not "fields" in the sense of having different data than "Descriptive Metadata". Those are In Memory, realtime, copies of the "Descriptive Metadata" SBF field and are there to overcome one limitation of Drupal 8:
+>Each Field can have a single "Formatter" setup per field. 
+
+But we want to re-usue the JSON data to show a Viewer, Show Metadata as HTML directly on the ADO/NODE landing page, and we want also to, for example, format sometimes images as Thumbnails and not using a IIIF viewer only. This CopyFields (Legal term) have also a nice Performance advantage. Drupal needs to fetch only once the data from the real Field, "Descriptive Metadata", from the database. And ...
```
```suggestion
But we want to re-usue the JSON data to show a Viewer, Show Metadata as HTML directly on the ADO/NODE landing page, and we want also to, for example, format sometimes images as Thumbnails and not using a IIIF viewer only. This CopyFields (Legal term) have also a nice Performance advantage. Drupal needs to fetch only once the data from the real Field, "Descriptive Metadata", from the database. And then just makes the data available in real time to its copies. That makes all fast, very very fast! And of course flexible. As you dig more into Archipelago you will see the benefits of this approach. Finally, if you need to, you can make more CopyFields. But the reality is, there is a single, only one, SBF in each Digital Object and it is named "Descriptive Metadata". 
```

marlo-longley commented:
@DiegoPino this is very helpful writing. I am grateful because it will make my life easier! I didn't find many errors at all but tried to suggest a few things. Take them or leave them! Even without changes people will understand what you say.

DiegoPino committed:
Update docs/webformsasinput.md

Better wording from @marlo

Co-Authored-By: Marlo Longley <marlo-longley@users.noreply.github.com>

DiegoPino committed:
Update docs/webformsasinput.md

add "This is" what lives ...

Co-Authored-By: Marlo Longley <marlo-longley@users.noreply.github.com>

DiegoPino committed:
Address Marlo's review

@marlo-longley i think i covered everything. Will give this a second pass and them merge. Thanks for all the good corrections

DiegoPino commented:
@marlo-longley i think i addressed all your comments and your review. Please let me know if i forgot something. I will merge once i have your confirmation. Thanks, all your comments made so much sense and it reads way better now.

marlo-longley commented on docs/webformsasinput.md:
```diff
@@ -0,0 +1,161 @@
+- Which Widgets are available will depend on the "type" of field the Content Entity has. 
+  - Example: A `Node` Title will have a single Text Input with some options, like the size of the Textfield used to feed it.
+  - More Complex and fun fields, like the ones of type `SBF` (strawberryfield), will provide a larger list of possible Widgets, ranging from raw JSON input (which you could select if your data was already in the right format) to the reason we are reading this: `Webform driven Widgets`. These Widgets include:
+   - ones the _webform_strawberrfield_ Drupal module provides
```
```suggestion
   - ones the _webform_strawberryfield_ Drupal module provides
```

marlo-longley replied:
also, why is it _webform not webform? just curious.

DiegoPino replied:
uhh.. totally right! Good eyes 👀

DiegoPino replied:
you mean the cursive/italic/emphasis _cursive_ = `_cursive_`

marlo-longley replied:
Ah my bad

DiegoPino replied:
All good! MD code reviews are the worst, because of unexpected/invisible styling. Thanks!

marlo-longley commented:
@DiegoPino found one small type. Otherwise great. I am hoping, in the end, we can simplify this document when the UI is also simpler!! But now it is so valuable.

DiegoPino committed:
Update docs/webformsasinput.md

Typo in webform_strawberryfield

Co-Authored-By: Marlo Longley <marlo-longley@users.noreply.github.com>
